### PR TITLE
types: migrate test-infra to testify for binary_literal_test.go

### LINKED ISSUE #27994: migrate test-infra to testify for `binary_literal_test.go`

### PR DESCRIPTION
**Pull request by hanlins**
Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

### What problem does this PR solve?

Issue Number: close #27994 

### Release note 

```release-note
NONE
```

### CODE REVIEW

File: types/binary_literal_test.go [hunk 1]
```diff
@@ -15,271 +15,269 @@
...
+	"testing"
+
 	"github.com/pingcap/tidb/sessionctx/stmtctx"
-	"github.com/pingcap/tidb/util/testleak"
+	"github.com/stretchr/testify/require"
 )
 
-var _ = Suite(&testBinaryLiteralSuite{})
+func TestBinaryLiteral(t *testing.T) {
+	t.Run("TestTrimLeadingZeroBytes", func(t *testing.T) {
```
Comment (xhebox) [inline types/binary_literal_test.go:26]: Sub-tests could be paralleld like `t.Parallel()`.

File: types/binary_literal_test.go [hunk 2]
```diff
@@ -15,271 +15,269 @@
...
+			{[]byte{0x1, 0x0}, []byte{0x1, 0x0}},
+			{[]byte{0x0, 0x1}, []byte{0x1}},
+			{[]byte{0x0, 0x0, 0x0}, []byte{0x0}},
+			{[]byte{0x1, 0x0, 0x0}, []byte{0x1, 0x0, 0x0}},
+			{[]byte{0x0, 0x1, 0x0, 0x0, 0x1, 0x0, 0x0}, []byte{0x1, 0x0, 0x0, 0x1, 0x0, 0x0}},
+			{[]byte{0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0x1, 0x0, 0x0}, []byte{0x1, 0x0, 0x0, 0x1, 0x0, 0x0}},
+		}
+		for _, item := range tbl {
+			b := trimLeadingZeroBytes(item.Input)
+			require.Equal(t, b, item.Expected, fmt.Sprintf("%#v", item))
```
Comment (xhebox) [inline types/binary_literal_test.go:43]: For testify, it is `Equal(t, exected, got)`. So all your `require.Equal` got reversed arguments order.

File: types/binary_literal_test.go [hunk 3]
```diff
@@ -15,271 +15,269 @@
...
-		hex, err := ParseHexStr(t.Input)
-		if t.IsError {
-			c.Assert(err, NotNil, Commentf("%#v", t))
-		} else {
-			c.Assert(err, IsNil, Commentf("%#v", t))
-			c.Assert([]byte(hex), DeepEquals, t.Expected, Commentf("%#v", t))
+		for _, item := range tbl {
+			hex, err := ParseHexStr(item.Input)
+			if item.IsError {
+				require.NotNil(t, err, fmt.Sprintf("%#v", item))
```
Comment (xhebox) [inline types/binary_literal_test.go:138]: Could use `NoError(..)` or `Error(..)` for better semantic.

**Review by xhebox — COMMENTED** [review]
Overall LGTM, I'll compare the diff more carefully after addressing feedbacks.

File: types/binary_literal_test.go (hunk 2 above)
Comment (hanlins) [reply types/binary_literal_test.go]: Done

File: types/binary_literal_test.go (hunk 3 above)
Comment (hanlins) [reply types/binary_literal_test.go]: Done, PTAL

**Review by xhebox — APPROVED** [review]

**Review by tisonkun — COMMENTED** [review]
Thanks for your contribution @hanlins !

I'd like to know why do you want to turn all these tests in subtest? We don't have to do it.

**Comment by hanlins** [comment]
Hi @tisonkun, thanks for the review! I intended to group the tests in the same test but different subtests so people know these tests are related, and it's easy to identify result for each test from the output, something like this:
```
    --- PASS: TestBinaryLiteral/TestCompare (0.00s)
    --- PASS: TestBinaryLiteral/TestToBitLiteralString (0.00s)
    --- PASS: TestBinaryLiteral/TestTrimLeadingZeroBytes (0.00s)
    --- PASS: TestBinaryLiteral/TestString (0.00s)
    --- PASS: TestBinaryLiteral/TestParseHexStr#01 (0.00s)
    --- PASS: TestBinaryLiteral/TestParseHexStr (0.00s)
    --- PASS: TestBinaryLiteral/TestToString (0.00s)
    --- PASS: TestBinaryLiteral/TestNewBinaryLiteralFromUint (0.00s)
    --- PASS: TestBinaryLiteral/TestParseBitStr#01 (0.00s)
    --- PASS: TestBinaryLiteral/TestToInt (0.00s)
    --- PASS: TestBinaryLiteral/TestParseBitStr (0.00s)
```

File: types/binary_literal_test.go [hunk 4]
```diff
@@ -15,271 +15,280 @@
...
+			{"b'1101000011001010110110001101100011011110010000001110111011011110111001001101100011001000010000001100110011011110110111100100000011000100110000101110010'", []byte("hello world foo bar"), false},
+			{"B'1101000011001010110110001101100011011110010000001110111011011110111001001101100011001000010000001100110011011110110111100100000011000100110000101110010'", []byte("hello world foo bar"), false},
+			{"0b01101000011001010110110001101100011011110010000001110111011011110111001001101100011001000010000001100110011011110110111100100000011000100110000101110010", []byte("hello world foo bar"), false},
+			{"b'01101000011001010110110001101100011011110010000001110111011011110111001001101100011001000010000001100110011011110110111100100000011000100110000101110010'", []byte("hello world foo bar"), false},
+			{"B'01101000011001010110110001101100011011110010000001110111011011110111001001101100011001000010000001100110011011110110111100100000011000100110000101110010'", []byte("hello world foo bar"), false},
+		}
+		for _, item := range tbl {
+			b, err := ParseBitStr(item.Input)
+			if item.IsError {
+				require.NotNil(t, err, fmt.Sprintf("%#v", item))
```
Comment (tisonkun) [inline types/binary_literal_test.go:102]:
```suggestion
				require.Error(t, err)
```

ditto others. testify already provides us rendered error message.

File: types/binary_literal_test.go [hunk 5]
```diff
@@ -15,271 +15,280 @@
...
+			} else {
+				require.Nil(t, err, fmt.Sprintf("%#v", item))
```
Comment (tisonkun) [inline types/binary_literal_test.go:104]:
```suggestion
				require.NoError(t, err, fmt.Sprintf("%#v", item))
```

ditto others.

**Comment by tisonkun** [comment]
> Hi @tisonkun, thanks for the review! I intended to group the tests in the same test but different subtests so people know these tests are related, and it's easy to identify result for each test from the output, something like this:

Ok. Make sense.

File: types/binary_literal_test.go (hunk 4 above)
Comment (hanlins) [reply types/binary_literal_test.go]: Done

File: types/binary_literal_test.go (hunk 5 above)
Comment (hanlins) [reply types/binary_literal_test.go]: Done, PTAL

**Review by tisonkun — APPROVED** [review]

**Pull request merged**